cloud/storage/core/libs/common/dynamic_persistent_table.h

```diff
@@ -0,0 +1,631 @@
+        {
+            ++Index;
+            SkipEmptyRecords();
+            return *this;
+        }
+
+        TIterator operator++(int)
+        {
+            auto tmp = *this;
+            ++*this;
+            return tmp;
+        }
+
+        TStringBuf operator*()
+        {
+            return Table.GetRecordWithValidation(Index);
+        }
+
+        TStringBuf operator->()
+        {
+            return Table.GetRecordWithValidation(Index);
+        }
+
+        ui64 GetIndex() const
+        {
+            return Index;
+        }
+
+        ui64 GetDataSize() const
+        {
+            if (Index >= Table.MaxRecords) {
+                return 0;
+            }
+            return Table.DescriptorsPtr[Index].DataSize;
+        }
+
+    private:
+        void SkipEmptyRecords()
+        {
+            while (Index < Table.MaxRecords &&
+                   Table.DescriptorsPtr[Index].State != ERecordState::Stored)
+            {
+                ++Index;
+            }
+        }
+
+    private:
+        const TDynamicPersistentTable& Table;
+        ui64 Index;
+    };
+
+    TIterator begin() const
+    {
+        return TIterator(*this, 0);
+    }
+
+    TIterator end() const
+    {
+        return TIterator(*this, MaxRecords);
+    }
+
+public:
+    TDynamicPersistentTable(
+        const TString& fileName,
+        ui64 maxRecords,
+        ui64 initialDataAreaSize,
+        ui64 gapSpacePercentageCompactionThreshold)
+        : FileName(fileName)
+        , MaxRecords(maxRecords)
+        , InitialDataAreaSize(initialDataAreaSize)
+        , GapSpacePercentageCompactionThreshold(
+              gapSpacePercentageCompactionThreshold)
+        , DataAreaSize(initialDataAreaSize)
+    {
+        Init();
+    }
+
+    H* HeaderData()
+    {
+        return &HeaderPtr->Data;
+    }
+
+    TStringBuf GetRecord(ui64 index) const
+    {
+        return GetRecord(index, EDataRetrievalMode::NoValidation);
+    }
+
+    TStringBuf GetRecordWithValidation(ui64 index) const
+    {
+        return GetRecord(index, EDataRetrievalMode::WithValidation);
+    }
+
+    ui64 AllocRecord(ui64 dataSize)
+    {
+        ui64 index = InvalidIndex;
+        if (dataSize == 0) {
+            return InvalidIndex;
+        }
+
+        if (!FreeRecordIndexes.empty()) {
+            index = FreeRecordIndexes.front();
+            FreeRecordIndexes.pop_front();
+        } else if (NextFreeRecordIndex < MaxRecords) {
+            index = NextFreeRecordIndex++;
+        }
+
+        if (index == InvalidIndex) {
+            return InvalidIndex;
+        }
+
+        if (NextDataOffset + dataSize > DataAreaSize) {
+            TryCompactDataArea(dataSize);
+            if (NextDataOffset + dataSize > DataAreaSize) {
+                ExpandDataArea(dataSize);
+            }
+        }
+
+        DescriptorsPtr[index].DataOffset = NextDataOffset;
+        DescriptorsPtr[index].DataSize = dataSize;
+        DescriptorsPtr[index].State = ERecordState::Allocated;
```

Comment:
you can not mark entry allocated before Prev/NextDataIndex is update and consistent otherwise you can crush and have inconsistent desscriptor